gateway_tests/process/test_subscriptions.py

```diff
@@ -0,0 +1,298 @@
+import copy
+import functools
+import logging
+import math
+import time
+from typing import Any
+
+import pytest
+from epics import ca, dbr
+
+from .. import conftest
+
+logger = logging.getLogger(__name__)
+
+masks = pytest.mark.parametrize(
+    "mask",
+    [
+        pytest.param(dbr.DBE_VALUE, id="DBE_VALUE"),
+        pytest.param(dbr.DBE_LOG, id="DBE_LOG"),
+        pytest.param(dbr.DBE_ALARM, id="DBE_ALARM"),
+        pytest.param(dbr.DBE_PROPERTY, id="DBE_PROPERTY"),
+        pytest.param(dbr.DBE_VALUE | dbr.DBE_PROPERTY, id="DBE_VALUE|DBE_PROPERTY"),
+        pytest.param(dbr.DBE_VALUE | dbr.DBE_LOG, id="DBE_VALUE|DBE_LOG"),
+        pytest.param(dbr.DBE_VALUE | dbr.DBE_ALARM, id="DBE_VALUE|DBE_ALARM"),
+    ]
+)
+
+value_masks = pytest.mark.parametrize(
+    "mask",
+    [
+        pytest.param(dbr.DBE_VALUE, id="DBE_VALUE"),
+        pytest.param(dbr.DBE_VALUE | dbr.DBE_PROPERTY, id="DBE_VALUE|DBE_PROPERTY"),
+        pytest.param(dbr.DBE_VALUE | dbr.DBE_LOG, id="DBE_VALUE|DBE_LOG"),
+        pytest.param(dbr.DBE_VALUE | dbr.DBE_ALARM, id="DBE_VALUE|DBE_ALARM"),
+    ]
+)
+
+
+forms = pytest.mark.parametrize(
+    "form",
+    [
+        "ctrl",
+        "time",
+    ]
+)
+
+
+@pytest.mark.parametrize(
+    "pvname",
+    [
+        "HUGO:AI",
+        "HUGO:ENUM",
+        # "auto",      # <-- updates based on auto:cnt
+        # "auto:cnt",  # <-- updates periodically
+        "enumtest",
+        "gwcachetest",
+        "passive0",
+        "passiveADEL",
+        "passiveADELALRM",
+        "passiveALRM",
+        "passiveMBBI",
+        "passivelongin",
+        "bigpassivewaveform",
+        "fillingaai",
+        "fillingaao",
+        "fillingcompress",
+        "fillingwaveform",
+        "passivewaveform",
+    ]
+)
+@masks
+@forms
+@conftest.standard_test_environment_decorator
+def test_subscription_on_connection(pvname: str, mask: int, form: str):
+    """
+    Basic subscription test.
+
+    For the provided pv name, mask, and form (ctrl/time), do we receive the same
+    subscription updates on connection to the gateway/IOC?
+    """
+    gateway_events = []
+    ioc_events = []
+
+    def on_change(event_list, pvname=None, chid=None, **kwargs):
+        event_list.append(copy.deepcopy(kwargs))
+
+    with conftest.ca_subscription_pair(
+        pvname,
+        ioc_callback=functools.partial(on_change, ioc_events),
+        gateway_callback=functools.partial(on_change, gateway_events),
+        form=form,
+        mask=mask,
+    ):
+        time.sleep(0.1)
+
+    compare_subscription_events(
+        gateway_events,
+        form,
+        ioc_events,
+        strict=False,
+        nan_strict=False,
+    )
+
+
+def is_acceptable_nan_difference(value1, value2) -> bool:
+    """Are (value1, value2) either NaN or 0.0?"""
```

Comment:
Should this one also have an associated warning? I'm not sure, it's probably fine.

Reply:
I think it does in the mess that is `compare_subscription_events`
`nan_strict` set -> raise, unset -> warn